File: SparkXAdSDKDemo/SparkXAdDemo/Native/NativeAdViewController.m
//
//  NativeAdViewController.m
//  SparkXAdDemo
//
//  Created by Wei yang on 2021/2/4.
//  Copyright © 2021 SparkX. All rights reserved.
//

#import "NativeAdViewController.h"
#import <SparkXAdSDK/SparkXNativeExpressAd.h>

@interface NativeAdViewController ()<SparkXAdNativeExpressAdDelegate>
@property (nonatomic, strong) UILabel *statusLabel;

@property (nonatomic ,strong) UIView *nativeView;

@property (strong, nonatomic) SparkXNativeExpressAd *nativeAd;

@end

@implementation NativeAdViewController

- (UIView *)nativeView {
    if (!_nativeView) {
        _nativeView = [[UIView alloc]initWithFrame:CGRectMake(0, 200, [UIScreen mainScreen].bounds.size.width - 0, ([UIScreen mainScreen].bounds.size.width - 0)*0.3)];
        [self.view addSubview:_nativeView];
    }
    return _nativeView;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    self.view.backgroundColor = [UIColor whiteColor];
    self.edgesForExtendedLayout = UIRectEdgeNone;
    
    _statusLabel = [[UILabel alloc]init];
    [_statusLabel setFont:[UIFont systemFontOfSize:16]];
    [_statusLabel setTextColor:[UIColor redColor]];
    [_statusLabel setTextAlignment:NSTextAlignmentCenter];
    _statusLabel.translatesAutoresizingMaskIntoConstraints = NO;
    _statusLabel.text = @"Tap button to load and show Ad";
    [self.view addSubview:_statusLabel];
    
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-20-[_statusLabel]-20-|" options:0 metrics:nil views:NSDictionaryOfVariableBindings(_statusLabel)]];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-35-[_statusLabel(25)]" options:0 metrics:nil views:NSDictionaryOfVariableBindings(_statusLabel)]];
    
    
    UIButton *loadNativeAd = [UIButton buttonWithType:UIButtonTypeSystem];
    loadNativeAd.layer.borderWidth = 0.5;
    loadNativeAd.layer.cornerRadius = 8;
    loadNativeAd.layer.borderColor = [UIColor lightGrayColor].CGColor;
    loadNativeAd.translatesAutoresizingMaskIntoConstraints = NO;
    [loadNativeAd addTarget:self action:@selector(loadNativeAd:) forControlEvents:UIControlEventTouchUpInside];
    [loadNativeAd setTitle:@"Load Native Ad" forState:UIControlStateNormal];
    [loadNativeAd setTitleColor:[UIColor redColor] forState:UIControlStateNormal];
    [self.view addSubview:loadNativeAd];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-30-[loadNativeAd]-170-|" options:NSLayoutFormatAlignAllCenterX metrics:nil views:NSDictionaryOfVariableBindings(loadNativeAd)]];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:[_statusLabel]-20-[loadNativeAd(40)]" options:0 metrics:nil views:NSDictionaryOfVariableBindings(_statusLabel,loadNativeAd)]];
    
    UIButton *removeNativeAd = [UIButton buttonWithType:UIButtonTypeSystem];
    removeNativeAd.layer.borderWidth = 0.5;
    removeNativeAd.layer.cornerRadius = 8;
    removeNativeAd.layer.borderColor = [UIColor lightGrayColor].CGColor;
    removeNativeAd.translatesAutoresizingMaskIntoConstraints = NO;
    [removeNativeAd addTarget:self action:@selector(removeNativeAd:) forControlEvents:UIControlEventTouchUpInside];
    [removeNativeAd setTitle:@"removeNativeAd" forState:UIControlStateNormal];
    [removeNativeAd setTitleColor:[UIColor redColor] forState:UIControlStateNormal];
    [self.view addSubview:removeNativeAd];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-30-[removeNativeAd]-170-|" options:NSLayoutFormatAlignAllCenterX metrics:nil views:NSDictionaryOfVariableBindings(removeNativeAd)]];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:[loadNativeAd]-20-[removeNativeAd(40)]" options:0 metrics:nil views:NSDictionaryOfVariableBindings(loadNativeAd,removeNativeAd)]];
}

- (void)removeNativeAd:(id)sender {
    for (SparkXNativeExpressAdView *view in self.nativeView.subviews) {
        if ([view isKindOfClass:[SparkXNativeExpressAdView class]]) {
            [view removeFromSuperview];
        }
    }
    
}

- (void)loadNativeAd:(id)sender {
    
    self.nativeAd = [[SparkXNativeExpressAd alloc]initWithPlacementId:@"546FF24D-08C6-91AB-FF0F-9C930521E406-4" adSize:self.nativeView.frame.size];
    self.nativeAd.delegate = self;
    [self.nativeAd loadAd];
    
//    [SparkXAdNativeExpressAdManager instance].delegate = self;
//    [[SparkXAdNativeExpressAdManager instance]showAdFromRootView:self.nativeView rootViewController:self];
    self.statusLabel.text = @"Loading......";
}

#pragma mark----delegate

/**
 * 拉取原生模板广告成功
 */
- (void)nativeExpressAdSuccessToLoad:(SparkXNativeExpressAd *)nativeExpressAd views:(NSArray<__kindof SparkXNativeExpressAdView *> *)views {
    self.statusLabel.text = @"Ad loaded Success";
    SparkXNativeExpressAdView *view = views.firstObject;
    view.controller = self;
    [view render];
    [self.nativeView addSubview:view];
    view.frame = self.nativeView.bounds;
    
}


/**
 * 拉取原生模板广告失败
 */
- (void)nativeExpressAdFailToLoad:(SparkXNativeExpressAd *)nativeExpressAd error:(NSError *)error {
    self.statusLabel.text = @"Ad loaded fail";
}


/**
 * 原生模板广告渲染成功, 此时的 nativeExpressAdView.size.height 根据 size.width 完成了动态更新。
 */
- (void)nativeExpressAdViewRenderSuccess:(SparkXNativeExpressAdView *)nativeExpressAdView {
    self.statusLabel.text = @"Ad loaded render success";
}

/**
 * 原生模板广告渲染失败
 */
- (void)nativeExpressAdViewRenderFail:(SparkXNativeExpressAdView *)nativeExpressAdView {
    self.statusLabel.text = @"Ad loaded render fail";
}


/**
 * 原生模板广告曝光回调
 */
- (void)nativeExpressAdViewExposure:(SparkXNativeExpressAdView *)nativeExpressAdView {
    self.statusLabel.text = @"Ad loaded exposure";
}


/**
 * 原生模板广告被关闭
 */
- (void)nativeExpressAdViewClosed:(SparkXNativeExpressAdView *)nativeExpressAdView {
    
}

- (void)nativeExpressAdWillLeaveApplication:(SparkXNativeExpressAdView *)nativeExpressAdView {
    
}

@end

File: SparkXAdDemo-Swift/SparkXAdDemo_Swift/MainPage/NativeAdViewController.swift
//
//  NativeAdViewController.swift
//  SparkXAdDemo-Swift
//
//  Created by Wei yang on 2021/3/16.
//

import UIKit

class NativeAdViewController: UIViewController ,SparkXAdNativeExpressAdDelegate{
    
    private var _nativeAd:SparkXNativeExpressAd?;
    private lazy var _statusLabel = UILabel();
    private lazy var _nativeView = UIView();

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        self.navigationItem.title = "Native"
        self.view.backgroundColor = .white;
        let safeTop = Double(self.navigationController?.view.safeAreaInsets.top ?? 0.0)
        _nativeView.frame = CGRect.init(x: 0, y: 280, width: self.view.frame.size.width, height: self.view.frame.size.width * 0.3)
        self.view.addSubview(_nativeView)
        
       
        _statusLabel.font = UIFont.init(name: "PingFang-SC", size: 16)
        _statusLabel.textColor = UIColor.red
        _statusLabel.textAlignment = .center
        _statusLabel.translatesAutoresizingMaskIntoConstraints = false
        _statusLabel.text = "Tap button to load and show Ad"
        self.view.addSubview(_statusLabel);
        self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-20-[_statusLabel]-20-|", options: NSLayoutConstraint.FormatOptions.init(), metrics: nil, views: ["_statusLabel":_statusLabel]))
        self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|-top-[_statusLabel(25)]", options: NSLayoutConstraint.FormatOptions.init(), metrics: ["top":NSNumber.init(value:safeTop+100.0)], views: ["_statusLabel":_statusLabel]))
        
        let loadNativeAd = UIButton(type:.custom)
        loadNativeAd.layer.borderWidth = 0.5;
        loadNativeAd.layer.cornerRadius = 8;
        loadNativeAd.layer.borderColor = UIColor.lightGray.cgColor
        loadNativeAd.translatesAutoresizingMaskIntoConstraints = false
        loadNativeAd.addTarget(self, action: #selector(self.loadNativeAd), for: .touchUpInside)
        loadNativeAd.setTitle("Load Native Ad", for: .normal)
        loadNativeAd.setTitleColor(UIColor.red, for: .normal)
        self.view.addSubview(loadNativeAd)
        self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-30-[loadNativeAd]-170-|", options: NSLayoutConstraint.FormatOptions.init(), metrics: nil, views: ["loadNativeAd":loadNativeAd]))
        self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:[_statusLabel]-20-[loadNativeAd(40)]", options: NSLayoutConstraint.FormatOptions.init(), metrics: nil, views: ["_statusLabel":_statusLabel,"loadNativeAd":loadNativeAd]))
        
        let removeNativeAd = UIButton(type:.custom)
        removeNativeAd.layer.borderWidth = 0.5;
        removeNativeAd.layer.cornerRadius = 8;
        removeNativeAd.layer.borderColor = UIColor.lightGray.cgColor
        removeNativeAd.translatesAutoresizingMaskIntoConstraints = false
        removeNativeAd.addTarget(self, action: #selector(self.removeNativeAd), for: .touchUpInside)
        removeNativeAd.setTitle("removeNativeAd", for: .normal)
        removeNativeAd.setTitleColor(UIColor.red, for: .normal)
        self.view.addSubview(removeNativeAd)
        self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-30-[removeNativeAd]-170-|", options: NSLayoutConstraint.FormatOptions.init(), metrics: nil, views: ["removeNativeAd":removeNativeAd]))
        self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:[loadNativeAd]-20-[removeNativeAd(40)]", options: NSLayoutConstraint.FormatOptions.init(), metrics: nil, views: ["loadNativeAd":loadNativeAd,"removeNativeAd":removeNativeAd]))
    }
    

    @objc func loadNativeAd(_ sender:UIButton) -> Void {
        _nativeAd = SparkXNativeExpressAd.init(placementId: "546FF24D-08C6-91AB-FF0F-9C930521E406-4", adSize: _nativeView.frame.size)
        _nativeAd?.delegate = self
        _nativeAd?.load()
        
        _statusLabel.text = "Loading......"
        
    }
    
    @objc func removeNativeAd(_ sender:UIButton) -> Void {
        
        for obj in _nativeView.subviews {
            if obj.isKind(of: SparkXNativeExpressAdView.self) {
                obj.removeFromSuperview()
       
            }
        }
    }
    
    
    //MARK:----delegate
    func nativeExpressAdViewClosed(_ nativeExpressAdView: SparkXNativeExpressAdView) {
        
    }
    
    func nativeExpressAdViewClicked(_ nativeExpressAdView: SparkXNativeExpressAdView) {
        
    }
    
    func nativeExpressAdViewExposure(_ nativeExpressAdView: SparkXNativeExpressAdView) {
        _statusLabel.text = "Ad loaded exposure";
    }
    
    func nativeExpressAdViewRenderFail(_ nativeExpressAdView: SparkXNativeExpressAdView) {
        _statusLabel.text = "Ad loaded render fail";
    }
    
    func nativeExpressAdFail(toLoad nativeExpressAd: SparkXNativeExpressAd, error: Error) {
        _statusLabel.text = "Ad loaded fail";
    }
    
    func nativeExpressAdViewRenderSuccess(_ nativeExpressAdView: SparkXNativeExpressAdView) {
        _statusLabel.text = "Ad loaded render success";
    }
    
    func nativeExpressAdSuccess(toLoad nativeExpressAd: SparkXNativeExpressAd, views: [SparkXNativeExpressAdView]) {
        _statusLabel.text = "Ad loaded Success"
        let view:SparkXNativeExpressAdView = views[0]
        view.controller = self
        view .render()
        _nativeView .addSubview(view)
        view.frame = _nativeView.bounds
        
    }
    
    
    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}
